force-app/main/default/classes/CON_ContactMerge_CTRL.cls

```diff
@@ -950,98 +950,137 @@ public with sharing class CON_ContactMerge_CTRL {
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, label.npe01.Contact_Merge_Error_Merge_Failed + ' Error: ' + e));
+
+            return null;
+        }
+
+        return new PageReference('/' + winningContact.id);
+    }
+
+    // Find the winningContact record based the selected value of the winningContact FieldRow
+    private Contact getWinningContact() {
         for (FieldRow row : fieldRows) {
             if (row.showRadio && row.fieldName.equals(MASTER_KEY)) {
-                master = new Contact(id = row.selectedValue);
-                break;
+                return new Contact(Id = row.selectedValue);
             }
         }
+        
+        return null;
+    }
 
-        if (master != null) {
-            // Update the field values of master record based on the selected value for each field.
-            for (FieldRow row : fieldRows) {
-                if (row.showRadio && 
-                    !row.fieldName.equals(MASTER_KEY) && 
-                    row.selectedValue != master.id) {
-                    SObject selectedRecord = this.selectedRecords.get(row.selectedValue);
-
-                    // Sobject.put is not happy when sobject.get returns null. It throws an 
-                    // exception System.SObjectException: Illegal assignment from Object to String.
-                    // But instead when you pass a null directly to put, it works fine. 
-                    // And hence, this if statement.
-                    if (selectedRecord.get(row.fieldName) == null) {
-                        master.put(row.fieldName, null);
-                    } else {
-                        Object val = selectedRecord.get(row.fieldName);
-                        master.put(row.fieldName, val);
-                    }
-                }
+    private List<Contact> groupLosingContacts(Contact winningContact) {
+        List<Contact> losingContacts = new List<Contact>();
+        for (Contact c : this.selectedRecords.values()) {
+            if (c.Id != winningContact.Id) {
+                losingContacts.add(c);
             }
+        }
 
-            // Group all the loosing records separately.
-            List<Contact> losers = new List<Contact>();
+        return losingContacts;
+    }
 
-            for (Contact c : this.selectedRecords.values()) {
-                if (c.id != master.id) {
-                    losers.add(c);
-                }
+    // Update the field values of master record based on the selected value for each field.
+    private void populateContactWithNonUniqueFields(Contact winningContact) {
+        for (FieldRow row : fieldRows) {
+            if (row.showRadio
+                && !row.fieldName.equals(MASTER_KEY)
+                && row.selectedValue != winningContact.Id
+                && !UTIL_Describe.isFieldUnique('Contact', row.fieldName))
+            {
+                SObject selectedRecord = this.selectedRecords.get(row.selectedValue);
```

Comment:
it looks like there's duplicate code where with the same code block in `populateContactWithOnlyUniqueFields()`. Could we extract this logic to a method to call in both places?